### PR TITLE
u256: log, log2 and sqrt impl (#5329)

### LINKED ISSUE #5329: `math.sw` primitive `u256` missing trait implementations
No

- `sqrt`
- `logarithm`
- `binary logarithm`

### PR DESCRIPTION
**Pull request by sudhackar**
## Description
simple changes that uses newton's method for sqrt and decomposition to `u64` for `log2` - `log` builds up on `log2` for `u256`

This should fix #5329

Ping @SwayStar123 
## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by sudhackar** [comment]
Sorry for closing the old PR - it closed when I rolled back the 2 commits automatically.

File: sway-lib-std/src/math.sw
Comment (SwayStar123) [inline sway-lib-std/src/math.sw:10]: Trait implementation functions are supposed to show the documentation written where the trait functions are defined, not a separate definition for every implementation, as such please move this into the function as a comment (2 slashes instead of 3)

File: sway-lib-std/src/math.sw
Comment (SwayStar123) [inline sway-lib-std/src/math.sw:242]: Please move this implementation above the implementation for u64

File: sway-lib-std/src/math.sw
Comment (SwayStar123) [inline sway-lib-std/src/math.sw:214]: Please move this implementation above the implementation for u64

File: sway-lib-std/src/math.sw
Comment (SwayStar123) [inline sway-lib-std/src/math.sw:215]: Similarly to previous comment regarding documentation, this documentation can be removed

**Review by SwayStar123 — DISMISSED** [review]
Left some comments regarding structure/documentation, but I am unfamiliar with the math/algorithms for the operations, so I have requested someone experienced with that to also review.

Thank you for the PR!

File: sway-lib-std/src/math.sw
Comment (sudhackar) [reply sway-lib-std/src/math.sw:242]: Same as the other one - `log` depends on `log2` for `u256` and hence should be after `log2`
```
error
   --> /home/sudhackar/personal/sway/sway-lib-std/src/math.sw:152:30
    |
150 |
151 |     fn log(self, base: Self) -> Self {
152 |         let self_log2 = self.log2();
    |                              ^^^^ No method named "log2" found for type "u256".
153 |         let base_log2 = base.log2();
154 |         self_log2 / base_log2
    |
____

```

File: sway-lib-std/src/math.sw
Comment (xunilrj) [inline sway-lib-std/src/math.sw:16]: What happens when `x0` is zero? We are not testing this case below, right?

File: sway-lib-std/src/math.sw
Comment (IGI-111) [reply sway-lib-std/src/math.sw:214]: @tritao Is this particular linearity example something we should have fixed already or something we need the collection context for?

Comment (sudhackar) [reply sway-lib-std/src/math.sw:214]: Are we waiting on #5572 to close before we merge this then?

**Review by xunilrj — APPROVED** [review]

**Pull request merged**
